File: DocumentAnalysisSystem/Models/Document.cs
using System.Collections.Generic;
using DocumentAnalysisSystem.Visitors;

namespace DocumentAnalysisSystem.Models
{
    public class Document : DocumentComponent
    {
        private readonly List<DocumentComponent> _children = new();

        public Document(string name) : base(name)
        {
        }

        public override IReadOnlyList<DocumentComponent> Children => _children.AsReadOnly();

        public void Add(DocumentComponent component)
        {
            _children.Add(component);
        }

        public override void Accept(IDocumentVisitor visitor)
        {
            visitor.Visit(this);
            foreach (var child in _children)
            {
                child.Accept(visitor);
            }
        }
    }
}

File: DocumentAnalysisSystemJava/src/main/java/com/documentanalysis/models/Document.java
package com.documentanalysis.models;

import com.documentanalysis.visitors.IDocumentVisitor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Document extends DocumentComponent {
    private final List<DocumentComponent> children = new ArrayList<>();

    public Document(String name) {
        super(name);
    }

    @Override
    public List<DocumentComponent> getChildren() {
        return Collections.unmodifiableList(children);
    }

    public void add(DocumentComponent component) {
        children.add(component);
    }

    @Override
    public void accept(IDocumentVisitor visitor) {
        visitor.visit(this);
        for (DocumentComponent child : children) {
            child.accept(visitor);
        }
    }
}
